### ISSUE TITLE
Autosubmit DB problems

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#510]
In GitLab by @pablohe on Feb 18, 2020, 16:28

I'm getting errors in the Autosubmit DB

```sh
> ./create_testcases.py --des weekly
Traceback (most recent call last):
  File "./create_testcases.py", line 26, in <module>
    main(args)
  File "./create_testcases.py", line 13, in main
    cursor.execute(query)
sqlite3.OperationalError: disk I/O error

```
If I execute it again, works fine

When I try to delete:

```sh
> autosubmit delete -f t059

Removing experiment directory...

Deleting experiment from database...
Autosubmit's database version is -1. Current version is 1. Updating...
[CRITICAL] Can not update database: table experiment_backup already exists
[ERROR] Connection to database could not be established: Database version could not be updated
Experiment t059 deleted
```
If I try again it works.

deleting another experiment I get:

```sh
[CRITICAL] Unhandled exception on Autosubmit: disk I/O error
Traceback (most recent call last):
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/autosubmit.py", line 422, in parse_args
    return Autosubmit.delete(args.expid, args.force)
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/autosubmit.py", line 663, in delete
    return Autosubmit._delete_expid(expid)
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/autosubmit.py", line 512, in _delete_expid
    ret = delete_experiment(expid_delete)
  File "/shared/earth/software/autosubmit/3.12.0b-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0b0-py2.7.egg/autosubmit/database/db_common.py", line 295, in delete_experiment
    'WHERE name=:name', {'name': experiment_id})
OperationalError: disk I/O error

```

@kserradell @mcastril @dbeltrankyl @wuruchi @ jberlin

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 18, 2020, 16:30

I did not updates today, 

`[ERROR] Connection to database could not be established: Database version could not be updated`

not sure what this means

**Comment by kinow** [comment]
In GitLab by @pablohe on Feb 18, 2020, 16:38

I found a journal file if it helps

```
bscearth315 /esarchive/autosubmit  > lrt ecearth.db*
-rw-rw-rw-+ 1 pechevar Earth 141068288 feb 18 16:37 ecearth.db
-rw-rw-rw-+ 1 eadmin   Earth      4640 feb 18 16:37 ecearth.db-journal
```

**Comment by kinow** [comment]
In GitLab by @ jberlin on Feb 18, 2020, 16:39

Hi,
I having the same issue , but occurs randomly:

```
3.12.0b0
autosubmit version: 
Traceback (most recent call last):
  File "./testing_suite.py", line 383, in <module>
    main()
  File "./testing_suite.py", line 313, in main
    my_experiments.append(Experiment_Factory.create(my_config.model, expid, my_config.settings))
  File "/home/Earth/jberlin/Development/testing_suite/source/experiment/experiment_factory.py", line 13, in create
    return Ecearth_Experiment(expid, my_config)
  File "/home/Earth/jberlin/Development/testing_suite/source/experiment/ecearth_experiment.py", line 9, in __init__
    super(Ecearth_Experiment, self).__init__(expid, my_config)
  File "/home/Earth/jberlin/Development/testing_suite/source/experiment/experiment.py", line 31, in __init__
    self.description=self.description()
  File "/home/Earth/jberlin/Development/testing_suite/source/experiment/experiment.py", line 64, in description
    cursor.execute("SELECT description FROM listexp WHERE name=\"%s\";" % self.id)
sqlite3.OperationalError: disk I/O error

```

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 18, 2020, 16:42

I'm contacting sysadmin.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 18, 2020, 16:45

http://sqlite.1065341.n5.nabble.com/sqlite3-OperationalError-disk-I-O-error-td18568.html

Based on the info of that web, it seems that the issue is with the journal not being able to create fast enough and should only be an issue for expid, delete and other write operations.

**Comment by kinow** [comment]
In GitLab by @pablohe on Feb 18, 2020, 16:59

mentioned in issue auto-ecearth3#1072

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 18, 2020, 17:48

Is that issue solved?

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 18, 2020, 17:49

Sysadmin did not find anything wrong in the server.

**Comment by kinow** [comment]
In GitLab by @pablohe on Feb 18, 2020, 18:01

Hi Now is working fine. Thanks.

**Comment by kinow** [comment]
In GitLab by @pablohe on Feb 18, 2020, 18:08

closed

**Comment by kinow** [comment]
In GitLab by @pablohe on Feb 19, 2020, 10:02

Hi @wuruchi 

Wich DB are you talking about?

Did you take a look of this wiki: https://earth.bsc.es/gitlab/es/auto-ecearth3/wikis/web_experiments

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 19, 2020, 10:04

Hi @pablohe 

ecearth.db

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 19, 2020, 10:40

According to my log, around Tue, 18 Feb 2020 17:02:05 +0100 (CET) something happened that killed job_times. So, job_times is no more and cannot be recovered unless a full database dump is restored.

Anyway, I will proceed to create a new database that we will store the table job_times. That's the bulk of the size of ecearth.db. Once I am done with those changes, I will let you know so you can attempt to delete job_times (which will prove very hard to do) in ecearth.db to free that space, although having that corrupted table there will not affect any other operation. 

The main reason I am doing this is that since job_times is continuously updated, any copy and paste of ecearth.db will inevitably corrupt this table. Indeed, it should have been done in a separate database since the beginning, or even in a more reliable database platform. That is also the case of experiment_status, which is potentially updated every 5 minutes; however, with a lower probability.

In conclusion, the new experiment tables that were created with the purpose of providing Autosubmit GUI with more information will be moved out of ecearth.db.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 19, 2020, 10:47

Thanks @wuruchi . However, even if you write every 5 minutes, it should not have been corrupted. 

So we wait for the copy to delete the table. I guess we can do it even if it is corrupted.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 19, 2020, 11:51

Hi @mcastril 

One of the downsides of using sqlite3 is that if you make a copy of the .db file while an insert or update is being performed, that copy will have a corrupted version of the table that was being modified. That is one explanation of what happened to job_times, although it might as well be another cause. However, this situation signals that the experiment's data that is collected by the API should be in its own database.

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 19, 2020, 12:27

If there is any issue, remember we backup the database daily at 1:00AM.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 19, 2020, 12:41

I would like access to the last backup.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 19, 2020, 13:07

@kserradell Do not worry, I found another restore point.

**Comment by kinow** [comment]
In GitLab by @pablohe on Feb 19, 2020, 13:16

@wuruchi are you changing the production DB right?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 19, 2020, 13:18

Hi @pablohe 

I am populating a new database as_times.db. Then, I will deploy the updated API to bscesweb04.

As soon as that is done I will let you know so you can free the unused space from job_times in ecearth.db.

I am not applying any more modifications or updates to ecearth.db.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 19, 2020, 14:43

Hi @pablohe 

I have finished deploying the updated API and the workers that populate the database queried by Autosubmit API. Feel free to delete job_times and all its content from ecearth.db.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 19, 2020, 16:41

Thanks @wuruchi . I can make a copy of ecearth.db an try to delete the table rightafter.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 19, 2020, 16:43

> One of the downsides of using sqlite3 is that if you make a copy of the .db file while an insert or update is being performed, that copy will have a corrupted version of the table that was being modified. 

I guess this is why you asked if someone did any DB restoring operation. Because if not this would not explain why it was corrupted.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 19, 2020, 17:06

@wuruchi I see that as_times.db has the same tables as ecearth.db . Are you not going to read different tables from each database? If not, how you ensure coherency?

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 19, 2020, 17:27

We'd better talk and decide which tables remain in each database. I also think it is better to maintain the more static, core information in ecearth.db , which is the production one and more sensible.

Extra functionality for times, dependencies caching can be in the auxiliar database.

I will not touch anything for now.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 20, 2020, 09:51

Hello @mcastril 

That has been accounted for in the following way:

There are only 3 new tables that were added as part of the API information retrieval effort: experiment_status, experiment_times, job_times. Those are going to be continuously updated in `as_times.db`, while the other (original) tables were migrated just for consistency purposes and might as well be deleted from `as_times.db`, although that will not change the size of the database due to various reasons, the main one is that job_times is 99.9% of the database size. The API reads all information from `ecearth.db` except for the 3 tables previously mentioned, meaning that the experiment information is still read from `ecearth.db`, only the information about experiment status and completion times is read from `as_times.db`.

Currently, `job_times` in `as_times.db` has 1596087 rows, but most queries stay in the range of milliseconds or less than a second thanks to some basic usage of indexes.

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 20, 2020, 10:31

reopened

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 20, 2020, 10:31

I'm having the issue again

```
[CRITICAL] Unhandled exception on Autosubmit: attempt to write a readonly database
Traceback (most recent call last):
  File "/shared/earth/easybuild/opensuse-leap/15.1/x86_64/software/autosubmit/3.11.0-hotfix-foss-2019b-Python-2.7.16/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/autosubmit.py", line 399, in parse_args
    args.expand_status, args.notransitive)
  File "/shared/earth/easybuild/opensuse-leap/15.1/x86_64/software/autosubmit/3.11.0-hotfix-foss-2019b-Python-2.7.16/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/autosubmit.py", line 1964, in create
    "job_packages_" + expid).reset_table()
  File "/shared/earth/easybuild/opensuse-leap/15.1/x86_64/software/autosubmit/3.11.0-hotfix-foss-2019b-Python-2.7.16/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/job/job_package_persistence.py", line 34, in __init__
    self.db_manager.create_table(self.JOB_PACKAGES_TABLE, self.TABLE_FIELDS)
  File "/shared/earth/easybuild/opensuse-leap/15.1/x86_64/software/autosubmit/3.11.0-hotfix-foss-2019b-Python-2.7.16/lib/python2.7/site-packages/autosubmit-3.11.0_hotfix-py2.7.egg/autosubmit/database/db_manager.py", line 54, in create_table
    cursor.execute(create_command)
OperationalError: attempt to write a readonly database
```

This is a major issue. We should fix ASAP.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 20, 2020, 10:39

> Those are going to be continuously updated in `as_times.db`, while the other (original) tables were migrated just for consistency purposes and might as well be deleted from `as_times.db` 

Thanks @wuruchi , that is what I was thinking yesterday. I think it is better to delete them and keep only in each database the tables that will be updated.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 20, 2020, 10:41

Not sure if anybody touch something but for me it's working at this moment

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 20, 2020, 10:42

I suggest making an sql dump of `ecearth.db` not including experiment_status, experiment_times, job_times; and the views: listexpstatus, listexpstatusjob. Then create another sqlite3 database and import all the information from the sql dump. That would result in a new consistent database of a greatly reduced size.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 20, 2020, 10:46

@kserradell were you using the correct EXPID ? This error was familiar to me (sometimes I have used a wrong EXPID):

https://autosubmit.readthedocs.io/en/latest/faq.html#critical-unhandled-exception-on-autosubmit-attempt-to-write-a-readonly-database

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 20, 2020, 10:52

a2lz, I just created half an hour ago.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 20, 2020, 10:54

@kserradell 

I think that the problem is that the job_packages has been created with eadmin user 

-rw-r--r--+ 1 eadmin   Earth 2048 Feb 20 10:28 job_packages_a2lz.db


Miriam had this issue last week although she wasn't using eadmin user. I also had this issue some months ago, not sure if it is the same case.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 20, 2020, 10:54

I see:

-rw-rw-r--+ 1 kserrade Earth  646 Feb 20 10:46 job_list_a2lz.pkl

-rw-r--r--+ 1 eadmin   Earth 2048 Feb 20 10:28 job_packages_a2lz.db

Curious about `eadmin`.

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 20, 2020, 10:58

I had to say I'm using the new AS installation... Why this file is created with eadmin??? :confused:

**Comment by kinow** [comment]
In GitLab by @kserradell on Feb 20, 2020, 11:00

I've deleted the file and now is working. @dbeltrankyl and @pablohe we should check this issue when we test AS in the new image.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 20, 2020, 11:00

I don't know but it also happen with the actual image and with all AS versions apparently I had this issue with master branch, miriam with 3.12.0b and you with 3.11-0

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 20, 2020, 15:16

I checked the dump and it only takes some seconds, so it could be maybe done at any moment.

But it should be done all at once in a programmatic way so it takes only that, some seconds.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 24, 2020, 17:32

@wuruchi , could you do the backup + dump + restore?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 25, 2020, 09:22

Hello @mcastril 

Sure, I have done it before. I can do it around 17:30 today, or it can be performed early (around 08:30) tomorrow. You tell me.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Feb 27, 2020, 13:19

Hello @mcastril 

I completed the dump + restore yesterday early in the morning. Everything seems to be going just fine. The old database is still there under the name `ecearth_old.db`.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 2, 2020, 12:45

Thanks @wuruchi , everything seems to be alright.

**Comment by kinow** [comment]
In GitLab by @mcastril on Mar 2, 2020, 12:46

closed

**Comment by kinow** [comment]
In GitLab by @pabretonniere on Mar 25, 2020, 18:52

The same issue (job_packages.db with no write access for me but only eadmin) just happened to me again.
I erased, relaunched and everything went fine. 

But I'll move the issue to the autosubmit project if that's OK, because I had a hard time finding it.

**Comment by kinow** [comment]
In GitLab by @pabretonniere on Mar 25, 2020, 18:52

moved from requests#1063
